be/src/storage/tablet_reader.cpp

```diff
@@ -47,13 +47,22 @@ void TabletReader::close() {
         _collect_iter.reset();
     }
     STLDeleteElements(&_predicate_free_list);
+    Rowset::release_readers(_rowsets);
     _obj_pool.clear();
 }
 
 Status TabletReader::prepare() {
     std::shared_lock l(_tablet->get_header_lock());
     auto st = _tablet->capture_consistent_rowsets(_version, &_rowsets);
+    if (!st.ok()) {
+        _rowsets.clear();
+        std::stringstream ss;
+        ss << "fail to init reader. tablet=" << _tablet->full_name() << "res=" << st;
+        LOG(WARNING) << ss.str();
+        return Status::InternalError(ss.str().c_str());
+    }
     _stats.rowsets_read_count += _rowsets.size();
+    Rowset::acquire_readers(_rowsets);
```

Comment:
Don't see corresponding release readers.
And I have one question, before this change the rowsets is not acquired, why does origin code work normally?

Reply:
Coreesponding release readers is in the `void TabletReader::close()`, this function is called by the `~TabletReader()`.
When TabletReader is destructed, it will automatically release the _rowsets.

Before this pr, ref counters for any rowsets are zero. So anyone who calls rowsets `close` function will free the rowsets. The origin code work normally because in the code we only `close` the rowsets when they are useless. For example, the txn writes the rowset is aborted...

The reason for that we need to add `acquire_readers` to increase the ref counter of rowsets here is follows. The stale rowsets may still used by other readers, we can only release those rowsets which are not used by any other readers. So we add `acquire_readers` here.